### PR TITLE
Implement NAdamW optimizer

### PR DESCRIPTION
**Pull request by janEbert**
NAdamW, which is simply NAdam with the AdamW weight decay term, has shown strong performance in optimizer comparisons such as
1. https://arxiv.org/abs/2211.09760
1. https://arxiv.org/abs/2306.07179

[The VeLO paper](https://arxiv.org/abs/2211.09760) argues its power lies in its ability to act as a superset of other popular optimizers.

This PR adds NAdamW by ~~copying and making very small adaptations to the NAdam implementation (just like AdamW and Adam). To see the small changes in better detail, you can `diff torch/optim/nadam.py torch/optim/nadamw.py`.~~ adding a boolean flag `decoupled_weight_decay` that activates NAdamW behavior (`False` by default) to NAdam.

Interest in the optimizer has also been shown in the PyTorch forums:
https://discuss.pytorch.org/t/nadamw-and-demon-optimizers/179778

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: torch/optim/nadamw.pyi [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from typing import Tuple
+
+from .optimizer import _params_t, Optimizer
+
+class NAdamW(Optimizer):
```
Comment (Skylion007) [inline torch/optim/nadamw.pyi:5]: There is no reason for it have it's own typing stub, just use the type annotations in the .py file

**Comment by albanD** [comment]
I'm not sure this is the best way for us to implement this (if we do want it). I would suggest you don't spend too much time on this PR until we have feedback from @janeyx99

**Comment by Skylion007** [comment]
Agreed, it's probably better to implement with inheritance / some modifications to the Nadam optim

**Comment by janEbert** [comment]
I appreciate the comments and commitment to code quality. I'm all for maybe just adding a boolean flag to NAdam to switch the weight decay mode.

However, please keep in mind that I took care for these changes to keep consistency with the rest of the code base. For example, NAdam also has a redundant typing stub, and AdamW is basically the same file as Adam except for the weight decay term in the update function and math description. I simply reproduced both of these for NAdamW.

**Comment by janeyx99** [comment]
Thanks for everyone's comments and for @janEbert's effort. The boolean flag would be a better way for sure here. BTW, we are attempting to merge the type stubs gradually, so it'd be good not to introduce more.

**Comment by janEbert** [comment]
Hey @janeyx99 thanks for giving the okay! The PR is much cleaner now. I did adapt the existing type stub to include the new flag.

For the functional `nadam` method, I added `decoupled_weight_decay` as a (not-only) kwarg with default value to keep backward compatibility. There is a small failure mode when people set `foreach` or `differentiable` via positional arguments, in which case the change is not backward-compatible.
However, it seemed those arguments are supposed to only be specified as kwargs and having them as positional arguments is only a workaround for TorchScript.

File: torch/optim/nadam.py [hunk 2]
```diff
@@ -119,8 +121,10 @@ def step(self, closure=None):
             &\rule{110mm}{0.4pt}                                                                 \\
             &\textbf{for} \: t=1 \: \textbf{to} \: \ldots \: \textbf{do}                         \\
             &\hspace{5mm}g_t           \leftarrow   \nabla_{\theta} f_t (\theta_{t-1})           \\
-            &\hspace{5mm}if \: \lambda \neq 0                                                    \\
+            &\hspace{5mm}\textbf{if} \: \lambda \neq 0 \text{ and not } \textit{decoupled_weight_decay} \\
```
Comment (janeyx99) [inline torch/optim/nadam.py:124]: Should mention decoupled_weight_decay as an input!

**Review by janeyx99 — COMMENTED** [review]
Thanks for the update! It looks good! Just one thing missing in the docs.

File: torch/optim/nadam.py (hunk 2 above)
Comment (janEbert) [reply torch/optim/nadam.py]: Thanks for noticing, done!

**Comment by janEbert** [comment]
Added the missing pseudocode input (and also fixed underscore escaping), PTAL. :)

**Review by janeyx99 — APPROVED** [review]
Thanks! Looks like it’ll need a rebase to resolve the merge conflicts.

**Comment by janEbert** [comment]
So we don't duplicate work, should I take care of that?

**Comment by janeyx99** [comment]
Yes, please update the PR! Otherwise, it won’t be able to land with merge conflicts.

**Comment by janEbert** [comment]
Rebased, PTAL.
